ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/query/BlockchainQueries.java

```diff
@@ -483,100 +487,34 @@ private TransactionWithMetadata transactionByHeaderAndIndex(
    */
   public List<LogWithMetadata> matchingLogs(
       final long fromBlockNumber, final long toBlockNumber, final LogsQuery query) {
-    if (fromBlockNumber > toBlockNumber || toBlockNumber > headBlockNumber()) {
-      return Lists.newArrayList();
-    }
-    List<LogWithMetadata> matchingLogs = Lists.newArrayList();
-    for (long blockNumber = fromBlockNumber; blockNumber <= toBlockNumber; blockNumber++) {
-      final Hash blockhash = blockchain.getBlockHashByNumber(blockNumber).get();
-      final boolean logHasBeenRemoved = !blockchain.blockIsOnCanonicalChain(blockhash);
-      final List<TransactionReceipt> receipts = blockchain.getTxReceipts(blockhash).get();
-      final List<Transaction> transaction =
-          blockchain.getBlockBody(blockhash).get().getTransactions();
-      matchingLogs =
-          generateLogWithMetadata(
-              receipts,
-              blockNumber,
-              query,
-              blockhash,
-              matchingLogs,
-              transaction,
-              logHasBeenRemoved);
+    try {
+      return LongStream.rangeClosed(fromBlockNumber, toBlockNumber)
+          .mapToObj(i -> matchingLogs(blockchain.getBlockHashByNumber(i).get(), query))
+          .flatMap(Collection::stream)
+          .collect(Collectors.toList());
+    } catch (NoSuchElementException nsee) {
+      return Collections.emptyList();
     }
-    return matchingLogs;
   }
 
-  public List<LogWithMetadata> matchingLogs(final Hash blockhash, final LogsQuery query) {
-    final List<LogWithMetadata> matchingLogs = Lists.newArrayList();
-    Optional<BlockHeader> blockHeader = blockchain.getBlockHeader(blockhash);
-    if (!blockHeader.isPresent()) {
-      return matchingLogs;
+  public List<LogWithMetadata> matchingLogs(final Hash blockHash, final LogsQuery query) {
```

Comment:
Would be nice to move the bulk of this logic (minus the query filtering) to `Blockchain` with method signature: `Optional<List<LogWithMetadata>> getLogsWithMetadata(BlockHash hash)`.  But, it may make sense to wait till we cut the "removed" field from `LogsWithMetadata`.

Reply:
Makes sense, I'll add that to the ticket about removing the `removed` field.

Reply:
Oh, it's already there. I think you had the same comment earlier